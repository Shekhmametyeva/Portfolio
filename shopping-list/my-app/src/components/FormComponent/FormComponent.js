import './FormComponent.css';
import React from 'react';
import ButtonSvgComponents from '../buttonSvgComponent/ButtonSvgComponent';
import { checkElement } from '../../api/api';

class FormComponent extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            value: '',
            load: false,
        }
    }
    sendData(value, data) {
        if(checkElement(value, data)) {
        this.setState({...this.state, load: true})
          fetch('http://localhost:5000/api/award', {
            method: 'POST',
            body: JSON.stringify({title: value}),
          }).then(() => {
            this.props.fetchData().then(() => {
                this.setState({...this.state, value: '', load: false})
            });
            
          })
        }  
    }
    handleSubmit(event){
        event.preventDefault();
        if(!this.state.load) {
            this.sendData(this.state.value, this.props.data)
        }
    }
    render() {
        const button = this.state.load ? <div className='loading'></div> : <ButtonSvgComponents type='submit' name={'add'} />
        return (
            <form className="shopping__form"  onSubmit={(event) => this.handleSubmit(event)}>
                <input
                readOnly={this.state.load}
                type='text' 
                placeholder='Что добавим в список' 
                value={this.state.value} 
                className='shopping__input' 
                onChange={(event) => {this.setState({...this.state, value: event.target.value})}}/>
                {button}
            </form>
            
        )
    }
}

export default FormComponent;