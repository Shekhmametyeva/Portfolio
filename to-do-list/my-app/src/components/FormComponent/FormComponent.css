.todo__form {
    width: 100%;
    padding: 10px 20px;
    background-color: #eee;
    border-radius: 16px;
    border: 2px solid transparent;
}
.todo__form:focus-within {
    outline: 2px solid #fff;
    border: 3px solid #016801;
}
.input__container {
    flex-grow: 1;
    display: flex;
}
.todo__input {
    width: 100%;
    font-size: 1.5rem;
    border: none;
    background-color: transparent;
}
.todo__input:focus {
    outline: none;
}
.valid {
    margin-bottom: -10px;
    padding-left: 20px;
    color: #076d078a;
}