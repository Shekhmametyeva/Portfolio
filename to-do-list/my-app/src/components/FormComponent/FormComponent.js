import './FormComponent.css';
import ButtonSvgComponent from '../ButtonSvgComponent/ButtonSvgComponent';
import React, { useState } from 'react';
import { checkValidity } from '../../helperFunctions/helperFunctions'
import RankFormComponent from '../RankFormComponent/RankFormComponent';

function FormComponent (props) {
    const [value, setValue] = useState('');
    const [valid, setValid] = useState(false);
    const [activeRank, setActiveRank] = useState('Личные');
    
    const prompt = valid ? <p className='valid'>{valid}</p> : null;
    return (
        <form   
            className='todo__form' 
            onSubmit={(event) => {
                event.preventDefault();
                const newValue = checkValidity(props, value, () => setValue(''), (text)=> setValid(text));
                if (newValue) {
                    props.editStateData(newValue, activeRank)
                } 
            }}> 
            <div className='input__container'>
                <input
                    autoFocus
                    readOnly={valid}
                    type='text' 
                    placeholder={valid ? '' : 'Что добавим в список?'} 
                    value={value} 
                    className='todo__input' 
                    onChange={(event) => {
                        setValue(event.target.value)
                    }}
                />
                <ButtonSvgComponent name='add' type='submit' disabled={valid}/>
            </div>
            {prompt}
            <RankFormComponent ranksList={props.ranksList} activeRank={activeRank} changeActiveRanc={(value) => setActiveRank(value)}/>
        </form>
    )
}

export default FormComponent