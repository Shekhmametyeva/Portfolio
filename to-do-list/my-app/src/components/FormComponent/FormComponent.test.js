import {render, screen, fireEvent} from '@testing-library/react';
import FormComponent from './FormComponent';

describe('тестируем FormComponent', () => {
    const ranksList = ['Личные', 'Семья', 'Работа', 'Покупки', 'Учеба'];
    test('render FormComponent', () => {
        render(<FormComponent ranksList={ranksList} />);
        const form = screen.getByTestId('form');
        const input = screen.getByPlaceholderText(/Что добавим в список?/i);
        expect(input).toBeInTheDocument();
        expect(form).toMatchSnapshot()
    })
    test('при пустом вводе плейсхолдер исчезает, появляется "поле не заполнено"', () => {
        render(<FormComponent ranksList={ranksList} />);
        const input = screen.getByPlaceholderText('Что добавим в список?');
        const btn = screen.getByTestId('btn-add-elem');
        fireEvent.input(input, {target: {value: ' '}});
        fireEvent.click(btn);
        expect(screen.getByText(/поле не заполнено/i)).toBeInTheDocument();
        expect(screen.queryByPlaceholderText('Что добавим в список?')).toBeNull()
    })
    test('инпут находится в фокусе при открытии и после отправки', () => {
        render(<FormComponent ranksList={ranksList} />);
        const input = screen.getByRole('textbox');
        expect(input).toHaveFocus()
        const btn = screen.getByTestId('btn-add-elem');
        fireEvent.click(btn);
        expect(screen.getByRole('textbox')).toHaveFocus()
    })
    test('инпут находится в фокусе после выбора категории', () => {
        render(<FormComponent ranksList={ranksList} />);
        const input = screen.getByRole('textbox');
        const rank = screen.getAllByRole('button');
        fireEvent.click(rank[4]);
        expect(input).toHaveFocus();
    })
    
})