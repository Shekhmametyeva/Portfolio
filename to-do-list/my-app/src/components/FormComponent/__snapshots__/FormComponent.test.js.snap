// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`тестируем FormComponent render FormComponent 1`] = `
<form
  class="todo__form"
  data-testid="form"
>
  <div
    class="input__container"
  >
    <input
      class="todo__input"
      placeholder="Что добавим в список?"
      type="text"
      value=""
    />
    <button
      class="button"
      data-testid="btn-add-elem"
      type="submit"
    >
      <svg
        class="icon"
      >
        <use
          xlink:href="sprite.svg#add"
        />
      </svg>
    </button>
  </div>
  <div
    class="form__button__container"
  >
    <div
      class="form__button active"
      data-testid="rank-button"
      value="Личные"
    >
      <button
        class="rank__button button"
        type="button"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="sprite.svg#Личные"
          />
        </svg>
      </button>
      <p>
        Личные
      </p>
    </div>
    <div
      class="form__button"
      data-testid="rank-button"
      value="Семья"
    >
      <button
        class="rank__button button"
        type="button"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="sprite.svg#Семья"
          />
        </svg>
      </button>
      <p>
        Семья
      </p>
    </div>
    <div
      class="form__button"
      data-testid="rank-button"
      value="Работа"
    >
      <button
        class="rank__button button"
        type="button"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="sprite.svg#Работа"
          />
        </svg>
      </button>
      <p>
        Работа
      </p>
    </div>
    <div
      class="form__button"
      data-testid="rank-button"
      value="Покупки"
    >
      <button
        class="rank__button button"
        type="button"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="sprite.svg#Покупки"
          />
        </svg>
      </button>
      <p>
        Покупки
      </p>
    </div>
    <div
      class="form__button"
      data-testid="rank-button"
      value="Учеба"
    >
      <button
        class="rank__button button"
        type="button"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="sprite.svg#Учеба"
          />
        </svg>
      </button>
      <p>
        Учеба
      </p>
    </div>
  </div>
</form>
`;
